### PR TITLE
Support scripting classmethod called with object instances

### PR DESCRIPTION
**Pull request by ppwwyyxx**
Currentlt classmethods are compiled the same way as methods - the first argument is self.
Adding a fake statement to assign the first argument to the class.
This is kind of hacky, but that's all it takes.

### CODE REVIEW

File: torch/jit/frontend.py [hunk 1]
```diff
@@ -217,7 +222,7 @@ def remove_prefix(text, prefix):
     return aligned_prefix + aligned_suffix
 
 
-def get_jit_def(fn, def_name, self_name=None):
+def get_jit_def(fn, def_name, self_name=None, is_classmethod=False):
```
Comment (SplitInfinity) [inline torch/jit/frontend.py:225]: Is it possible to determine if `fn` is a classmethod without passing an extra parameter? I think it is:
```
>>> class A:
...   @classmethod
...   def a(cls):
...     pass
...   def b(self):
...     pass
...   @staticmethod
...   def c(s, t):
...     pass
... 
>>> A.a.__self__
<class '__main__.A'>
>>> A.b.__self__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'function' object has no attribute '__self__'
>>> A.c.__self__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'function' object has no attribute '__self__'
```
So it seems that `__self__` is only present on classmethods? So I think we can do that check in `get_jit_def` directly without the extra parameter.

Comment (ppwwyyxx) [reply torch/jit/frontend.py:225]: Regular methods that are bounded to instances, e.g. `A().b` in your example will have `__self__`. Could `get_jit_def` be called with such methods?

Comment (SplitInfinity) [reply torch/jit/frontend.py:225]: From my knowledge, no. Can you try it to see if it's possible (to get rid of the extra function argument)?

Comment (ppwwyyxx) [reply torch/jit/frontend.py:225]: This fails many tests. I added `print(fn)` when `__self__` is present, and saw cases it is called with bound regular methods:
```
$python3 test/test_jit.py                                                                                                                                                          
Fail to import hypothesis in common_utils, tests are not derandomized                                                                                                                 
.................................<bound method BasicModule.forward of BasicModule()>                                                                                                  
E<bound method BasicModule.forward of BasicModule()>                                                                                                                                  
E<bound method BasicModule.forward of BasicModule()>                                                                                                                                  
E./private/home/yuxinwu/DL/pytorch/test/jit/test_builtins.py:127: DeprecationWarning: Please use assertEqual instead.                                                                 
  self.assertEquals(py_out, jit_out)                                                                                                                                                  
.<bound method TestBuiltins.test_has_attr.<locals>.Mod.forward of Mod(                                                                                                                
  (mods): ModuleList(                                                                                                                                                                 
    (0): HasA()                                                                                                                                                                       
    (1): HasB()                                                                                                                                                                       
  )                                                                                                                                                                                   
)>                                                                                                                                                                                    
<bound method ModuleList.forward of ModuleList(                                                                                                                                       
  (0): HasA()                                                                                                                                                                         
  (1): HasB()                                                                                                                                                                         
)>                                                                                                                                                                                    
E<bound method TestBuiltins.test_has_attr_invalid_args.<locals>.Mod.forward of Mod(                                                                                                   
  (mod): Linear(in_features=1, out_features=1, bias=True)                                                                                                                             
)>                                                                    
```

**Review by SplitInfinity — APPROVED** [review]

**Pull request closed without merging**